[Feature] CSV import and export support header

chenlinzhong opened the pull request:
# Proposed changes

Issue Number: open #8487

## Problem Summary:

- Add two new types to stream load boker load: csv_with_names and csv_with_name_sand_types
- Add two new types to export: csv_with_names and csv_with_names_and_types

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

chenlinzhong committed:
[Feature] CSV import and export support header

chenlinzhong committed:
[Feature] CSV import and export support header

chenlinzhong committed:
[Feature] CSV import and export support header,format code

chenlinzhong committed:
[Feature] CSV import and export support header,format code

cambyzju commented on be/src/exec/broker_scanner.cpp:
```diff
@@ -103,6 +103,10 @@ Status BrokerScanner::get_next(Tuple* tuple, MemPool* tuple_pool, bool* eof, boo
+            _skip_rows--;
+            continue;
+        }
         if (_skip_next_line) {
```
could we replace _skip_next_line by '_skip_rows=1' ?

chenlinzhong replied:
ok, the latest code has been replaced

cambyzju commented on be/src/exec/broker_scanner.cpp:
```diff
@@ -161,6 +193,11 @@ Status BrokerScanner::open_file_reader() {
         start_offset -= 1;
     }
+    //means first range 
+    if (start_offset == 0){
```
if start_offset=1, then in line 194, it will be set to start_offset=0.

it is still first range?

chenlinzhong replied:
Yes, the reason for subtracting 1 is that offset is plus 1 in the FE calculation (bytesperinstance = totalbytes / numinstances + 1), so for the first range  start_ Offset is always 0, and other range will not be 0

Gabriel39 committed:
[feature-wip](java-udf) support java UDF with fixed-length input and output (#8516)

This feature is propsoed in [DSIP-1](https://cwiki.apache.org/confluence/display/DORIS/DSIP-001%3A+Java+UDF). 
This PR support fixed-length input and output Java UDF. Phase I in DIP-1 is done after this PR.

To support Java UDF effeciently, I use no data copy in JNI call and all compute operations are off-heap in Java.
To achieve that, I use a UdfExecutor instead. 

For users, a UDF class must have a public evaluate method.

HappenLee committed:
[fix] Fix error crc32 method to cal uint128 and int128 (#8577)

liutang123 committed:
[fix] Fix Check_time return wrong value when exec show table status (#8578)

HappenLee committed:
[fix][vectorized] Fix bug of left semi/anti with other join conjunct (#8596)

cambyzju commented on be/src/http/action/stream_load.cpp:
```diff
@@ -98,6 +98,40 @@ static TFileFormatType::type parse_format(const std::string& format_str,
         if (compress_type.empty()) {
             format_type = TFileFormatType::FORMAT_JSON;
         }
+    } else if (boost::iequals(format_str, "CSV_WITH_NAMES")) {
```
looks these codes the same as boost::iequals(format_str, "CSV")？

chenlinzhong replied:
It has been optimized

cambyzju commented on be/src/http/action/stream_load.cpp:
```diff
@@ -98,6 +98,40 @@ static TFileFormatType::type parse_format(const std::string& format_str,
+    } else if (boost::iequals(format_str, "CSV_WITH_NAMES")) {
+        if (compress_type.empty()) {
+            format_type = TFileFormatType::FORMAT_CSVWITHNAMES_PLAIN;
+        }
```
add else?

chenlinzhong replied:
It has been optimized

cambyzju commented on be/src/http/action/stream_load.cpp:
```diff
@@ -558,6 +606,8 @@ Status StreamLoadAction::_process_put(HttpRequest* http_req, StreamLoadContext*
                      << ctx->brief();
         return plan_status;
     }
+    LOG(INFO) << "ThriftDebugString:" << apache::thrift::ThriftDebugString(ctx->put_result);
```
debug message use VLOG, and remove std::cout

chenlinzhong replied:
fixed

chenlinzhong committed:
[Feature] CSV import and export support header

chenlinzhong committed:
[Feature] CSV import and export support header

spaces-X committed:
[feature] Support pre-aggregation for quantile type (#8234)

Add a new column-type to speed up the approximation of quantiles.
1. The  new column-type is named `quantile_state` with fixed aggregation function `quantile_union`, which stores the intermediate results of pre-aggregated approximation calculations for quantiles.
2. support pre-aggregation of new column-type and quantile_state related functions.

BiteTheDDDDt committed:
[fix][vectorized] fix core on get_predicate_column_ptr && fix double copy on _read_columns_by_rowids (#8581)

morningman committed:
[fix](load) fix another bug that BE may crash when calling `mark_as_failed` (#8607)

Same as #8501

GoGoWen committed:
[doc] fix typo for session (#8610)

qidaye committed:
[doc] fix help module failed (#8617)

Introduced by #8509.
Docs title is duplicate.

BiteTheDDDDt committed:
[fix] fix core dump on deep_copy_tuple when data is null (#8620)

BiteTheDDDDt committed:
[Enhancement] add build paramnt ENABLE_JAVAUDF, BUILD_DOCS (#8612)

* add build parament ENABLE_JAVAUDF,BUILD_DOCS

HappenLee committed:
[fix] Fix coredump of stddev function (#8543)

This is only a temporary fix its performance is not ideal. Finally,
we need to reconstruct the functions of `stddev` and delete the interface of `insert_to_null_default ()`.

xinyiZzz committed:
[feature] (memory) Switch TLS mem tracker to separate more detailed memory usage (#8605)

In pr #8476, all memory usage of a process is recorded in the process mem tracker,
and all memory usage of a query is recorded in the query mem tracker,
and it is still necessary to manually call `transfer to` to track the cached memory size.

We hope to separate out more detailed memory usage based on Hook TCMalloc new/delete + TLS mem tracker.

In this pr, the more detailed mem tracker is switched to TLS, which automatically and accurately
counts more detailed memory usage than before.

caiconghui committed:
[refactor](mutex) Use std::mutex to replace Mutex and refactor some lock logic (#8452)

924060929 committed:
[test] support a lot of actions (#8632)

Support a lot of actions for regression testing framework.
e.g. thread, lazyCheck, onSuccess, connect, selectUnionAll, timer

Demo exists in ${DORIS_HOME}/regression-test/suites/demo

JusticeGong committed:
[Doris Manager][Doc]Basic User Documents for Doris Manager (#8609)

yangzhg committed:
[api-change] add soft limit of String type length (#8567)

1. add a config string_type_soft_limit to soft limit max length of string type
2. disable using String type in Key column, partition column and
   distribution column
3. remove String type alias BLOB for futrue use

dataroaring committed:
ow num is more accurate than column num in data_types (#8628)

Commit:
[Refactor][agent_task] Remove etl mgr and etl job pool from be (#8635)

zbtzbtzbt committed:
[Enhancement](load) speed up stream load for duplicate table, use template for faster get_type_info. (#8500)

Commit:
Revert "[Refactor][agent_task] Remove etl mgr and etl job pool from be (#8635)" (#8666)

This reverts commit 6bc982c37436acf288f566cf10e084731b80fa44.

chenlinzhong committed:
[Feature] CSV import and export support header

chenlinzhong committed:
[Feature] CSV import and export support header

chenlinzhong committed:
[Feature] CSV import and export support header

chenlinzhong committed:
[Feature] CSV import and export support header

BiteTheDDDDt committed:
[Refactor] Remove ununsed file (#8657)

yangzhg committed:
[chore] optimize aws thirdparty package download. (#8637)

yinzhijian committed:
[feature](balance) Support balance between disks on a single BE (#8553)

Current situation of Doris is that the cluster is balanced, but the disks of a backend may be unbalanced.
for example, backend A have two disks: disk1 and disk2, disk1's usage is 98%, but disk2's usage is only 40%.
disk1 is unable to take more data, therefore only one disk of backend A can take new data,
the available write throughput of backend A is only half of its ability, and we can not resolve this through load or 
partition rebalance now.

So we introduce disk rebalancer, disk rebalancer is different from other rebalancer(load or partition)
which take care of cluster-wide data balancing. it takes care about backend-wide data balancing.

[For more details see #8550](https://github.com/apache/incubator-doris/issues/8550)

whutpencil committed:
[fix](load) Fix null column bug in load's mapping column setting (#8625)

wangbo committed:
[fix](vectorization) Vectorization decimal arithmetic inconsistent (#8626)

adonis0147 committed:
[chore][be-test] Link gtest_main to provide default main function definition. (#8631)

airborne12 committed:
[fix](vec) fix coredump for aggregate function when delete large_data, due to alloc-dealloc-mismatch (#8641)

xinyiZzz committed:
[fix] fix memory leak in VDataStreamRecvr::SenderQueue (#8643)

After `VDataStreamRecvr::SenderQueue::close` clears `_block_queue`, calling 
`VDataStreamRecvr::SenderQueue::add_block` again will cause a memory leak.

So, change the lock position, like the other add_block and add_batch.

daikon12 committed:
[doc] optimize some doc expression (#8645)

emerkfu committed:
[doc] Update compilation.md (#8646)

Added solutions to the "fatal error: Killed signal terminated program ..."
problem encountered when compiling with Docker to the FAQ.

dataroaring committed:
[chore] add -rtlib=compiler-rt for UBSAN under clang (#8647)

morningman committed:
[fix](mini-load) Remove mini load in LOADING and PENDING state (#8649)

1. Remove some unused code.
2. handle mini load with wrong state
    1. For some historical reasons, some mini load jobs in LOADING state have not been cleared.
        As a result, new load jobs cannot be committed.
    2. If a mini load job is created right before FE restart, the mini load job will be in PENDING state forever.
        But it should be removed finally.

zxm17863626761 committed:
[doc] update export.md (#8650)

"where" should be in front of "to".

dataroaring committed:
[test] limit memory used by regression test framework (#8651)

adonis0147 committed:
[chore] Optimize build_lz4 in build-thirdparty.sh (#8653)

dataroaring committed:
[test] regression framework use RollingFileAppender by default (#8654)

zbtzbtzbt committed:
[doc] update doc of vec-execution-engine (#8655)

BiteTheDDDDt committed:
[doc] fix sql-mode document (#8662)

dataalive committed:
[doc] fix buffer pool default value (#8670)

morningman committed:
[community] Add more collaborators (#8672)

1. add dataroaring
2. remove qidaye because he has became the committer of Doris

discuss thread: https://lists.apache.org/thread/8bxnj7qw2p120v077nm8gny52m65d22r

xlwh committed:
[typo] Optimize some code comments (#8673)

yangzhg commented on gensrc/thrift/PlanNodes.thrift:
```diff
@@ -108,6 +108,22 @@ enum TFileFormatType {
     FORMAT_JSON,
     FORMAT_PROTO,
+    //csv withnames
+    FORMAT_CSVWITHNAMES_PLAIN,
```
may put header and compress type to another message is better

chenlinzhong replied:
Yes, the more reasonable way is to enumerate the file type and compression type separately. However, since the previous logic is that the two are put together, the expansibility is not very good. If you want to separate them, there are many changes involved and the cycle will be relatively long. Therefore, the previous way is still temporarily continued here

caiconghui committed:
[fix](ut) fix be ut failed (#8682)

chenlinzhong committed:
[Feature] CSV import and export support header,format

morningman committed:
[fix] fix compile error (#8688)

Introduced from PR #8643
The condition_variable can only wait unique_lock

dataroaring committed:
[fix] fix core dump when avg on not null decimal in empty table (#8681)

chenlinzhong committed:
[Feature] CSV import and export support header,rebase
